File: src/Sales/BrewUp.Sales.Entities/Entities/SalesOrder.cs
using BrewUp.Sales.SharedKernel.CustomTypes;
using BrewUp.Sales.SharedKernel.Messages.Events;
using BrewUp.Shared.Domain;
using BrewUp.Shared.ExternalContracts;

namespace BrewUp.Sales.Entities.Entities;

public class SalesOrder : BrewUpAggregateRoot
{
    public string SalesOrderNumber { get; private set; } = string.Empty;
    public DateTime SalesOrderDate { get; private set; }
    
    public string CustomerId { get; private set; } = string.Empty;
    public string CustomerName { get; private set; } = string.Empty;
    
    public DateTime SalesOrderDeliveryDate { get; private set; }
    
    public virtual ICollection<SalesOrderRow> SalesOrderRows { get; private set; } = [];
    
    public string Status { get; private set; } = string.Empty;
    
    protected SalesOrder()
    {}

    public static SalesOrder Create(SalesOrderId salesOrderId, SalesOrderNumber salesOrderNumber,
        SalesOrderDate salesOrderDate, CustomerId customerId, CustomerName customerName,
        SalesOrderDeliveryDate salesOrderDeliveryDate, IEnumerable<SalesOrderRowJson> rows)
    {
        return new SalesOrder(salesOrderId, salesOrderNumber, salesOrderDate, customerId, customerName,
            salesOrderDeliveryDate, rows);
    }

    private SalesOrder(SalesOrderId salesOrderId, SalesOrderNumber salesOrderNumber, SalesOrderDate salesOrderDate,
        CustomerId customerId, CustomerName customerName, SalesOrderDeliveryDate salesOrderDeliveryDate,
        IEnumerable<SalesOrderRowJson> rows)
    {
        var rowsArray = rows.ToArray();
        
        Id = salesOrderId.Value;
        SalesOrderNumber = salesOrderNumber.Value;
        SalesOrderDate = salesOrderDate.Value;
        CustomerId = customerId.Value;
        CustomerName = customerName.Value;
        SalesOrderDeliveryDate = salesOrderDeliveryDate.Value;
        SalesOrderRows = rowsArray.Select(row => SalesOrderRow.Create(new SalesOrderId(Id),
            new ProductId(row.ProductId), new ProductName(row.ProductName),
            row.Quantity, row.Price)).ToList();
        
        Status = "Created";
    
        RaiseEvent(new SalesOrderCreated(salesOrderId, salesOrderNumber, salesOrderDate, customerId, customerName, salesOrderDeliveryDate, rowsArray));
    }
}

File: src/Sales.FSharp/BrewUp.Sales.Entities.FSharp/Entities/SalesOrder.fs
namespace BrewUp.Sales.Entities.FSharp.Entities

open System
open BrewUp.Shared.FSharp.Domain
open BrewUp.Shared.FSharp.ExternalContracts
open BrewUp.Sales.SharedKernel.FSharp.CustomTypes
open BrewUp.Sales.SharedKernel.FSharp.Messages

type SalesOrder = {
    Id: string
    SalesOrderNumber: string
    SalesOrderDate: DateTime
    CustomerId: string
    CustomerName: string
    SalesOrderDeliveryDate: DateTime
    SalesOrderRows: SalesOrderRow list
    Status: string
    UncommittedEvents: obj list // Simplified events storage
}

module SalesOrder =
    let create (salesOrderId: SalesOrderId) (salesOrderNumber: SalesOrderNumber) (salesOrderDate: SalesOrderDate)
               (customerId: CustomerId) (customerName: CustomerName) (deliveryDate: SalesOrderDeliveryDate)
               (rows: SalesOrderRowJson array) (correlationId: Guid) =
        
        let orderRows = 
            rows
            |> Array.map (fun row ->
                SalesOrderRow.create 
                    salesOrderId
                    (ProductId.From row.ProductId)
                    (ProductName.Create row.ProductName)
                    row.Quantity
                    row.Price)
            |> Array.toList
        
        let salesOrder = {
            Id = salesOrderId.Value
            SalesOrderNumber = salesOrderNumber.Value
            SalesOrderDate = salesOrderDate.Value
            CustomerId = customerId.Value
            CustomerName = customerName.Value
            SalesOrderDeliveryDate = deliveryDate.Value
            SalesOrderRows = orderRows
            Status = "Created"
            UncommittedEvents = []
        }
        
        // Create domain event (simplified)
        let event = {
            SalesOrderId = salesOrder.Id
            SalesOrderNumber = salesOrder.SalesOrderNumber
            CustomerId = salesOrder.CustomerId
            CustomerName = salesOrder.CustomerName
            CreatedAt = DateTime.UtcNow
        }
        
        { salesOrder with UncommittedEvents = [event] }
    
    let close (salesOrder: SalesOrder) =
        let closedEvent = {
            SalesOrderId = salesOrder.Id
            ClosedAt = DateTime.UtcNow
        }
        
        { salesOrder with 
            Status = "Closed"
            UncommittedEvents = salesOrder.UncommittedEvents @ [closedEvent] }
    
    let markReadyForProcessing (salesOrder: SalesOrder) =
        let readyEvent = {
            SalesOrderId = salesOrder.Id
            SalesOrderNumber = salesOrder.SalesOrderNumber
            ProcessingRequested = DateTime.UtcNow
        }
        
        { salesOrder with 
            Status = "ReadyForProcessing"
            UncommittedEvents = salesOrder.UncommittedEvents @ [readyEvent] }